File: cpp/0004.median-of-two-sorted-arrays.cpp
class Solution
{
public:
    double findMedianSortedArrays(vector<int>& a, vector<int>& b)
    {
        int m = a.size();
        int n = b.size();
        if (!m && !n)
        {
            return 0.0;
        }
        bool need_another = false;
        if ((m + n) % 2 == 0)
        {
            need_another = true;
            int v;
            if (m && n)
            {
                v = max(a.back(), b.back());
            }
            else if (m)
            {
                v = a.back();
            }
            else if (n)
            {
                v = b.back();
            }
            n++;
            b.push_back(v);
        }
        int half = ((m + n) >> 1);
        int x = findFromFirstArray(a, b);
        if (x < 0)
        {
            swap(a, b);
            x = findFromFirstArray(a, b);
        }
        double res = a[x];
        if (need_another)
        {
            int y = half - x;
            int v;
            if (x - 1 >= 0 && y - 1 >= 0)
            {
                v = max(a[x - 1], b[y - 1]);
            }
            else if (x - 1 >= 0)
            {
                v = a[x - 1];
            }
            else if (y - 1 >= 0)
            {
                v = b[y - 1];
            }
            res = (res + v) / 2.0;
        }
        return res;
    }

private:
    int findFromFirstArray(vector<int>& a, vector<int>& b)
    {
        int m = a.size();
        int n = b.size();
        int half = ((m + n) >> 1);
        int st = 0;
        int ed = m - 1;
        while (st <= ed)
        {
            int x = ((st + ed) >> 1);
            int y = half - x;
            if (y < 0)
            {
                ed = x - 1;
                continue;
            }
            if (y > n)
            {
                st = x + 1;
                continue;
            }
            if (y > 0 && b[y - 1] > a[x])
            {
                st = x + 1;
                continue;
            }
            if (y < n && a[x] > b[y])
            {
                ed = x - 1;
                continue;
            }
            return x;
        }
        return -1;
    }
};

File: rust/0004_median-of-two-sorted-arrays.rs
impl Solution
{
    pub fn find_median_sorted_arrays(a: Vec<i32>, b: Vec<i32>) -> f64
    {
        let mut c = Vec::new();
        let m = a.len();
        let n = b.len();
        let all = m + n;
        let mut i = 0;
        let mut j = 0;

        while i < m && j < n
        {
            if (a[i] < b[j])
            {
                c.push(a[i]);
                i += 1;
            }
            else
            {
                c.push(b[j]);
                j += 1;
            }
        }

        while i < m
        {
            c.push(a[i]);
            i += 1;
        }

        while j < n
        {
            c.push(b[j]);
            j += 1;
        }

        if all % 2 == 0
        {
            return (c[all / 2 - 1] + c[all / 2]) as f64 / 2.0;
        }
        else
        {
            return c[all / 2] as f64;
        }
    }
}
